client/disputes/redirect-to-transaction-details.tsx

```diff
@@ -0,0 +1,51 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+
+/**
+ * Internal dependencies.
+ */
+import { useDispute } from 'data/index';
+import Page from 'components/page';
+import { Dispute } from 'wcpay/types/disputes';
+import { Charge } from 'wcpay/types/charges';
+import { getAdminUrl } from 'wcpay/utils';
+
+const RedirectToTransactionDetails = ( {
+	query: { id: disputeId },
+}: {
+	query: { id: string };
+} ): JSX.Element => {
+	const { dispute, isLoading } = useDispute( disputeId );
+	const disputeObject = dispute || ( {} as Dispute );
+	const disputeIsAvailable = ! isLoading && dispute && disputeObject.id;
+	// Why would dispute.charge ever be a string?
```

Comment:
Our interfaces for Disputes are not 100% accurate here.:
* When fetching a charge, `charge.dispute.charge: string` 
* When fetching a dispute, `dispute.charge: Charge`

Currently, we use `Dispute` as the interface for both of these response types.

I've had a first crack at solving this in #7311 – unsure if it's the right way to go.

Reply:
Thanks for the tip - will take a look!

Reply:
Added some comments. This feels a bit like typescript chicanery at first glance. Hoping there's a simple non-hack solution! If we need to start sketching out hierarchical types that implies maintenance risk IMO (object oriented vibes).

Reply:
I've updated my comment, since the type is correct in my view. The part I don't understand – is casting the correct approach an "OR" type like this? I.e. let me know if `as Charge` is a hack, or if I need a runtime type check.

Reply:
I see type-casting as a risk in this scenario since it makes assumptions about the code that may not always be true. E.g. we may in the future change the response type of this endpoint to always return `dispute.charge: string` and not realise that this has been overridden (`as Charge`) throughout the codebase.

I think a more robust way to assert that `dispute.charge === Charge` would be via a [type guard function](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) e.g.:

```ts
const isCharge = ( charge: Charge | string ): charge is Charge => {
	return typeof charge !== 'string';
};
const { charge } = dispute;
if ( isCharge( charge ) ) {
	// charge.payment_intent exists!
```

However, [since we know that `GET disputes/{id}` endpoint will return a `dispute` with `dispute.charge: Charge`](https://github.com/Automattic/woocommerce-payments/blob/1a76dcefde87b8ae6f8e2bfefa5dad01b76ddba5/client/data/disputes/resolvers.js#L52), it would be better to fix this upstream (the `Dispute` interface) since we have full control here.

I think it's OK to follow this up in a future PR, however.

Reply:
Thanks - I look forward to understanding this better in the upstream PR. 

This kind of thing feels very heavy compared to classic loose-duck-types JavaScript, so I want to really understand how to use TypeScript to strengthen things in a lean/elegant way.